### ISSUE TITLE
Compiling Nuklear as a library

### ISSUE THREAD

**Issue by BieHDC** [Immediate-Mode-UI/Nuklear#300]
I am trying to compile Nuklear as a standalone library that outputs a single .o file which i can include in the main.c file later. The goal is to reduce compile time, however after playing with the #define'es, i am not able to make this work.
I cant find any documentation on that ether and also no example.
Can someone please provide one?

Expectation how it should work like:
gcc -o nklib.o nuklear.h nuklear_backend.h ...
gcc -o example example.c nklib.o ...

**Comment by unspecd** [comment]
Try something like this:
```c
/* nuklear-defs.h */
#ifndef NUKLEAR_DEFS_H
#define NUKLEAR_DEFS_H

#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_FONT_BAKING
#define NK_INCLUDE_STANDARD_IO

#endif
```
```c
/* nuklear.c */
#include "nuklear-defs.h"
#define NK_IMPLEMENTATION
#include <nuklear.h>
```
```c
/* main.c */
#include "nuklear-defs.h"
#include <nuklear.h>

int main()
{
    /* nk_init() etc. */
    return 0;
}
```
```makefile
# Makefile

CFLAGS = -Ipath/to/nuklear

all: example

example: main.o nuklear.o
	$(CC) -o $@ $^

%.o: %.c
	$(CC) $(CFLAGS) -c -o $@ $<
```

**Comment by FrostKiwi** [comment]
>  library that outputs a single .o file 

Though almost the same, a .o file is not a library ;]

If you compile using a backend provided in src/demos, they accidentality used a Macro defined only for internal use. This is fixed here: https://github.com/Immediate-Mode-UI/Nuklear/pull/298

You can generate separate Nuklear.h and Nuklear.c files using the packer, but simpler and faster version would be:
When you need Nuklear's header, include all needed options, without NK_IMPLEMENTATION.
When you need the Objectfile, create an empty Nuklear.c file and paste the same options, include the NK_IMPLEMENTATION define and then the header again. Just be sure, NK_IMPLEMENTATION is never written twice. Done.
Should look like this:
```
/* This Nuklear C file that compiles to an object file */
#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_STANDARD_IO
#define NK_INCLUDE_STANDARD_VARARGS
#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
#define NK_INCLUDE_FONT_BAKING
#define NK_KEYSTATE_BASED_INPUT
#define NK_IMPLEMENTATION

#include "nuklear.h"
```

**Comment by BieHDC** [comment]
> > library that outputs a single .o file
> 
> Though almost the same, a .o file is not a library, but an object file ;]
> 
Yes, that is what i meant.

> If you compile using a backend provided in src/demos, it will produce an error if the Header if split from the library implementation. The demos accidentality used a Macro defined only for internal use. This is fixed here: #298
So that was the reason i had weird issues.
Thanks for your answer.

**Comment by RobLoach** [comment]
:rocket:
